### PR TITLE
Add bucket limit to GCS service

### PR DESCRIPTION
**Pull request by CuriousVini**
image

### CODE REVIEW

File: wrangler-service/src/main/java/co/cask/wrangler/service/gcs/GCSService.java [hunk 1]
```diff
@@ -154,8 +153,9 @@ private boolean validateConnection(String connectionId, Connection connection,
   @GET
   @Path("/connections/{connection-id}/gcs/explore")
   public void exploreGCS(HttpServiceRequest request, HttpServiceResponder responder,
-                              @PathParam("connection-id") final String connectionId,
-                              @QueryParam("path") String path) {
+                         @PathParam("connection-id") final String connectionId,
+                         @QueryParam("path") String path,
+                         @QueryParam("bucket-limit") @DefaultValue("1000") int bucketLimit) {
```
Comment (albertshau) [inline wrangler-service/src/main/java/co/cask/wrangler/service/gcs/GCSService.java]: didn't notice before, but I think this should be named 'limit' and not 'bucket-limit' for both endpoints

File: wrangler-service/src/main/java/co/cask/wrangler/service/gcs/GCSService.java [hunk 2]
```diff
@@ -510,12 +511,12 @@ public void specification(HttpServiceRequest request, final HttpServiceResponder
...
   }
 
-  private Collection<Bucket> getBuckets(Storage storage, Set<String> whitelist) {
+  private Collection<Bucket> getBuckets(Storage storage, Set<String> whitelist, int bucketLimit) {
     // this will include buckets that can be listed by the service account, but may not include all buckets
     // in the whitelist, if the whitelist contains publicly accessible buckets from other projects.
     // do some post-processing to filter out anything not in the whitelist and also try and lookup buckets
     // that are in the whitelist but not in the returned list
-    Page<Bucket> list = storage.list();
+    Page<Bucket> list = storage.list(Storage.BucketListOption.pageSize(bucketLimit));
```
Comment (albertshau) [inline wrangler-service/src/main/java/co/cask/wrangler/service/gcs/GCSService.java]: this sounds like it's a page size limit, not a limit for the entire response. list.iterateAll() will iterate through every page. There is another method that iterates through just the elements in the page.

Is there a limit for how big the requested page size can be? If not, we just need to change the iterateAll() to the other one. If there is, we need to go page by page and limit it ourselves.

File: wrangler-service/src/main/java/co/cask/wrangler/service/gcs/GCSService.java [hunk 3]
```diff
@@ -275,12 +280,19 @@ public void run(DatasetContext datasetContext) throws Exception {
           object.addProperty("wrangle", isWrangeable);
         }
         values.add(object);
+        if (values.size() >= bucketLimit) {
```
Comment (CuriousVini) [inline wrangler-service/src/main/java/co/cask/wrangler/service/gcs/GCSService.java]: @albertshau should we use a separate query param to limit number of objects in a bucket?

File: wrangler-service/src/main/java/co/cask/wrangler/service/gcs/GCSService.java [hunk 4]
```diff
@@ -154,8 +152,9 @@ private boolean validateConnection(String connectionId, Connection connection,
   @GET
   @Path("/connections/{connection-id}/gcs/explore")
   public void exploreGCS(HttpServiceRequest request, HttpServiceResponder responder,
-                              @PathParam("connection-id") final String connectionId,
-                              @QueryParam("path") String path) {
+                         @PathParam("connection-id") final String connectionId,
+                         @QueryParam("path") String path,
+                         @QueryParam("limit") @DefaultValue("1000") int bucketLimit) {
```
Comment (albertshau) [inline wrangler-service/src/main/java/co/cask/wrangler/service/gcs/GCSService.java]: nit: this is more like an objectLimit, as it returns more than just buckets.

**Review by albertshau — APPROVED** [review]
one nit about renaming a variable. fix and merge

File: wrangler-service/src/main/java/co/cask/wrangler/service/gcs/GCSService.java (hunk 3 above)
Comment (albertshau) [reply wrangler-service/src/main/java/co/cask/wrangler/service/gcs/GCSService.java]: no i think its fine to have a single limit, as it's a single API.

**Pull request merged**